src/main/java/org/breedinginsight/brapi/v2/services/BrAPITrialService.java

```diff
@@ -52,4 +93,278 @@ private long countGermplasm(UUID programId, String trialDbId) throws ApiExceptio
+                addObsVarDataToRow(row, obs, includeTimestamp, obsVars, program);
+                rowByOUId.put(ouId, row);
+            }
+        }
+    }
+
+    private String getOUId(BrAPIObservationUnit ou) {
+        BrAPIExternalReference ouXref = Utilities.getExternalReference(
+                        ou.getExternalReferences(),
+                        String.format("%s/%s", referenceSource, ExternalReferenceSource.OBSERVATION_UNITS.getName()))
+                .orElseThrow(() -> new RuntimeException("observation unit id not found"));
+        return ouXref.getReferenceID();
+    }
+
+    private void addObsVarDataToRow(
+            Map<String, Object> row,
+            BrAPIObservation obs,
+            boolean includeTimestamp,
+            List<BrAPIObservationVariable> obsVars,
+            Program program) {
+        // get observation variable for BrAPI observation
+        BrAPIObservationVariable var = obsVars.stream()
+                .filter(obsVar -> obs.getObservationVariableName().equals(obsVar.getObservationVariableName()))
+                .collect(Collectors.toList()).get(0);
+
+        String varName = Utilities.removeProgramKey(obs.getObservationVariableName(), program.getKey());
+        if (var.getScale().getDataType().equals(BrAPITraitDataType.ORDINAL)) {
+            row.put(varName, Integer.parseInt(obs.getValue()));
+        } else if (var.getScale().getDataType().equals(BrAPITraitDataType.NUMERICAL) ||
+                var.getScale().getDataType().equals(BrAPITraitDataType.DURATION)) {
+            row.put(varName, Double.parseDouble(obs.getValue()));
+        } else {
+            row.put(varName, obs.getValue());
+        }
+
+        if (includeTimestamp) {
+            row.put(String.format("TS:%s",varName), obs.getObservationTimeStamp());
+        }
+    }
+
+    public List<BrAPIObservationVariable> getDatasetObsVars(String datasetId, Program program) throws ApiException, DoesNotExistException {
+        List<BrAPIListSummary> lists = listDAO.getListByTypeAndExternalRef(
+                BrAPIListTypes.OBSERVATIONVARIABLES,
+                program.getId(),
+                String.format("%s/%s", referenceSource, ExternalReferenceSource.DATASET.getName()),
+                UUID.fromString(datasetId));
+        if (lists == null || lists.isEmpty()) {
+            throw new DoesNotExistException("Dataset observation variables list not returned from BrAPI service");
+        }
+        String listDbId = lists.get(0).getListDbId();
+        BrAPIListsSingleResponse list = listDAO.getListById(listDbId, program.getId());
+        List<String> obsVarNames = list.getResult().getData();
+        return obsVarDAO.getVariableByName(obsVarNames, program.getId());
+    }
+
+    public BrAPITrial getExperiment(Program program, UUID experimentId) throws ApiException {
+        List<BrAPITrial> experiments = trialDAO.getTrialsByExperimentIds(List.of(experimentId), program);
+        return experiments.get(0);
```

Comment:
May be worth having a check here to ensure the list returned has at least one element.  If it doesn't, throw an error.  This will prevent `ArrayIndexOutOfBounds` errors from occurring

Reply:
done

Reply:
done